### PR TITLE
STORM-2702: storm-loadgen

### PR DESCRIPTION
**Pull request by revans2**
Some tools to be able to generate load on a cluster for testing.

### CODE REVIEW

**Comment by roshannaik** [comment]
The latency reported in the TVL report differs from what is shown in the UI by a factor of 10k. Division error maybe ?

Here is a sample report indicating mean latency of 84,246ms but the UI for the same run shows 8.4ms.

start_time(s),end_time(s),completion_rate(tuple/s),mean(ms),99%ile(ms),99.9%ile(ms),cores,mem(MB),failed
0,30,73695.16666666667,1700.898775864534,2562.719743,2621.439999,1.6000666666666667,75.75511169433594,0
....
389,419,230711.53333333333,**84246**.15838892311,87308.632063,87375.740927,4.228833333333333,105.1185302734375,0

**Comment by roshannaik** [comment]
another observation of latency reported differing from UI and report (this time by ~120x):
`
start_time(s),end_time(s),completion_rate(tuple/s),mean(ms),99%ile(ms),99.9%ile(ms),cores,mem(MB),failed
....
181,211,145589.76666666666,99707.19604026663,107374.182399,107508.400127,4.9433333333333325,2169.0212020874023,0
211,241,145994.63333333333,114849.46619218022,122876.329983,122943.438847,4.990833333333334,1467.2594909667969,0
241,271,144041.5,130208.78991536044,138512.695295,138781.130751,4.888733333333334,1794.4858627319336,0
`
stats

**Comment by roshannaik** [comment]
Would be nice if the latencies are trimmed off at 1 digit after decimal point

**Comment by revans2** [comment]
@roshannaik The latency reported is a simulation of kafka or something like it.  The start time, is not when the message is emitted by the spout.  The start time is when the message would have been inserted into something like kafka to be processed.  So if the topology cannot keep up with the throughput the latency grows accordingly.  I should document it better in the README.

I made the decision not to cut the numbers off in CSV/TSV because those files are really intended to be read by a machine.  They are not really human readable anyways.

You can adjust the columns to what you want and ids is the one that spits out the topology ids that are running.

**Comment by revans2** [comment]
@roshannaik I just added in support for a new default reporter that writes the data out in a fixed width format that should be much more human readable, with formatting, and it is added as a reporter if nothing else is writing to stdout or stderr.

**Comment by revans2** [comment]
I think I am done with new features unless someone else comes up with something else that I missed out on. I really would appreciate some reviews. @knusbaum if you have the time I know you expressed interest in this patch.

@roshannaik if you really want metrics about the emit to ack/fail latency I can add those in too, but the only time that they become useful is if the topology cannot keep up with the load placed on it.

**Comment by knusbaum** [comment]
Partial review done.

File: examples/storm-loadgen/src/main/java/org/apache/storm/loadgen/InputStream.java [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+
+    public static class Builder {
+        private String fromComponent;
+        private String toComponent;
+        private String id;
+        private NormalDistStats execTime;
+        private NormalDistStats processTime;
+        private GroupingType groupingType = GroupingType.SHUFFLE;
+
+        public String getFromComponent() {
```
[the remark this answers is not in the document]
Comment (revans2) [reply examples/storm-loadgen/src/main/java/org/apache/storm/loadgen/InputStream.java]: No they don't but it was the most convenient way to pull out the needed information while still building the object.

**Comment by revans2** [comment]
@knusbaum I think I have addressed all of your review comments so far.

File: examples/storm-loadgen/src/main/java/org/apache/storm/loadgen/InputStream.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (knusbaum) [reply examples/storm-loadgen/src/main/java/org/apache/storm/loadgen/InputStream.java]: Ahh, sorry. I missed their use.

**Comment by roshannaik** [comment]
@revans2  sorry missed your msgs from here. 
- Yes would be nice to have the actual latency (as reported in UI) to also be included here. I also run TVL in modes where the topology cant keep up. i have been having to collect it separately. can report 0 latency if ACKer is disabled.
- I often copy and paste these csv results into excel for easy viewing. So the csv results with decimals limited to 3 digits would be nice. Unless the decimals after 3 digits is useful for machine readable not  that many decimal places even for machine readable
- One usability improvement would for the tool *tee* whatever it is printing into the report onto the console as well. helps in multiple ways.

**Comment by roshannaik** [comment]
Another useful thing would be to kill to topo with a simple Ctrl-C on the cmd line using a shutdown handler.

**Comment by revans2** [comment]
@roshannaik I think I have added in all of your feature requests so far.

**Comment by revans2** [comment]
@knusbaum do you have any other comments?

**Comment by knusbaum** [comment]
+1

**Comment by revans2** [comment]
@roshannaik I merged this into master, but if you want any changes to it please just ask and I'll see what I can do.

**Comment by roshannaik** [comment]
sure thanks.

**Pull request merged**